src/Storages/ObjectStorage/StorageObjectStorageSink.cpp

```diff
@@ -83,20 +83,21 @@ void StorageObjectStorageSink::finalize()
     {
         writer->finalize();
         writer->flush();
-        write_buf->finalize();
     }
     catch (...)
     {
         /// Stop ParallelFormattingOutputFormat correctly.
         release();
         throw;
     }
+
+    write_buf->finalize();
 }
 
 void StorageObjectStorageSink::release()
 {
     writer.reset();
-    write_buf.reset();
+    write_buf->finalize();
```

Comment:
I assume finalize() was not called here on purpose https://github.com/ClickHouse/ClickHouse/commit/2c018f5261553dd6106639f22c148fbdd61d8fc4#diff-a118e043f3175929dbe8a96c6dcc7db80e31ddae5581f29b4286eb9e2d9d925d

Reply:
I was confused because StorageFile and StorageURL had `finalize()` in the same place (and before StorageObject only StorageS3 didn't have `finalize()`). I am still not 100% sure that it's safe to just call `.reset()` here as we still can call finalize in destructor. Let me check the write buffers that could be used, maybe it's actually safe and we can use `.reset()` in all such places

Reply:
> and before StorageObject only StorageS3 didn't have finalize()

Yeah I also noticed that during refactoring (this finalize() was also there for `StorageAzure`) and changed it the same way, but it resulted in failing s3 tests